File: mobile/plugins/foray-tts/android/src/main/java/ai/jwlabs/foura/tts/KokoroOrtProbeEngine.java
package ai.jwlabs.foura.tts;

import android.content.Context;
import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.nio.LongBuffer;
import java.util.HashMap;
import java.util.Map;

import ai.onnxruntime.OnnxTensor;
import ai.onnxruntime.OrtEnvironment;
import ai.onnxruntime.OrtSession;

/**
 * K-01's runtime on Android: ONNX Runtime, driven directly.
 *
 * <p>The iOS half ({@code KokoroOrtProbeEngine.swift}) carries the full
 * argument for why this exists, why ORT rather than sherpa-onnx or MLX, and
 * why the samples are counted and thrown away. The two differ only where the
 * platforms do, and there are exactly two places:
 *
 * <ol>
 *   <li><b>The model has to leave the APK before it can be opened.</b> Android
 *       assets are DEFLATE-compressed in the package, so there is no file
 *       descriptor to memory-map and no path to hand ORT. The alternative to
 *       extracting is {@code createSession(byte[])}, which means an 82 MB
 *       array on the JVM heap — on a phone whose default heap may be 192 MB,
 *       that is a plausible OOM AND it would corrupt the one reading this card
 *       exists to take, because peak memory is a go/no-go clause. So the asset
 *       is copied once to {@code cacheDir} and opened by path, and the copy
 *       happens in {@link #create} BEFORE any stopwatch starts, so it is not
 *       counted as model-load time. iOS needs none of this: a bundle resource
 *       is already a file.</li>
 *   <li><b>The memory figure means something different.</b> The plugin reports
 *       {@code Debug.getNativeHeapAllocatedSize()} alongside the JVM figure,
 *       because ORT's arena is native. That is the plugin's business, not
 *       this class's, and is commented there.</li>
 * </ol>
 *
 * <p>NOT A NARRATION PATH. No {@code speak}, no {@code AudioTrack}, no
 * foreground service. It answers {@link ForayTtsPlugin.KokoroProbeEngine} —
 * ids in, timings out — and nothing in {@code queue-manager.js} can reach it.
 */
final class KokoroOrtProbeEngine implements ForayTtsPlugin.KokoroProbeEngine {

    private static final String TAG = "ForayTts/kokoro";

    /** Kokoro v1.0 emits 24 kHz. A property of the model card, not of the
     *  graph — a wrong value here would scale every RTF by the ratio without
     *  changing anything a reader could see. */
    private static final double SAMPLE_RATE = 24_000d;

    /** The style matrix: 510 token-lengths x 256 floats. 510 * 256 * 4 =
     *  522,240 bytes, which is exactly what {@code fetch-models.mjs} pins. */
    private static final int STYLE_ROWS = 510;
    private static final int STYLE_DIM = 256;

    static final String VOICE_ASSET = "af_heart.bin";

    private final String modelPath;
    private final float[] style;
    private OrtEnvironment env;
    private OrtSession session;
    private String provider = "cpu";

    private KokoroOrtProbeEngine(String modelPath, float[] style) {
        this.modelPath = modelPath;
        this.style = style;
    }

    /**
     * Build one, or {@code null} when anything needed is absent or malformed.
     *
     * <p>A FAILED CONSTRUCTION IS NOT A CRASH and not a zero: the plugin
     * reports {@code engine-absent}, one of the four closed reason codes a
     * founder reads off the screen and {@code player/kokoro-probe.js} knows
     * by name.
     */
    static KokoroOrtProbeEngine create(Context ctx) {
        if (ctx == null) return null;
        try {
            byte[] raw = readAsset(ctx, VOICE_ASSET);
            int expected = STYLE_ROWS * STYLE_DIM * 4;
            if (raw == null || raw.length != expected) {
                /* A voice file of the wrong length is a build that fetched
                   something else. Refusing beats synthesizing with 256 floats
                   read out of the middle of an unrelated file, which would
                   produce sound and therefore a number. */
                Log.e(TAG, "voice asset is " + (raw == null ? "absent" : raw.length + " bytes")
                        + ", expected " + expected);
                return null;
            }
            FloatBuffer fb = ByteBuffer.wrap(raw).order(ByteOrder.LITTLE_ENDIAN).asFloatBuffer();
            float[] style = new float[fb.remaining()];
            fb.get(style);

            File extracted = extractModel(ctx);
            if (extracted == null) return null;
            return new KokoroOrtProbeEngine(extracted.getAbsolutePath(), style);
        } catch (Throwable t) {
            Log.e(TAG, "could not prepare the Kokoro probe engine", t);
            return null;
        }
    }

    /** The asset, on disk, reused across runs. Re-extracted when the size does
     *  not match, which is the cheap half of "is this the same file" and the
     *  only half available without hashing 82 MB on a phone. */
    private static File extractModel(Context ctx) {
        try {
            File out = new File(ctx.getCacheDir(), ForayTtsPlugin.MODEL_ASSET);
            long assetSize = -1;
            try (InputStream probe = ctx.getAssets().open(ForayTtsPlugin.MODEL_ASSET)) {
                assetSize = probe.available();
            } catch (Throwable ignored) { /* available() is a hint; fall through */ }
            if (out.exists() && assetSize > 0 && out.length() == assetSize) return out;
            try (InputStream in = ctx.getAssets().open(ForayTtsPlugin.MODEL_ASSET);
                 OutputStream os = new FileOutputStream(out)) {
                byte[] buf = new byte[1 << 16];
                int n;
                while ((n = in.read(buf)) > 0) os.write(buf, 0, n);
            }
            return out;
        } catch (Throwable t) {
            Log.e(TAG, "could not extract the Kokoro model from assets", t);
            return null;
        }
    }

    private static byte[] readAsset(Context ctx, String name) {
        try (InputStream in = ctx.getAssets().open(name)) {
            java.io.ByteArrayOutputStream bos = new java.io.ByteArrayOutputStream();
            byte[] buf = new byte[1 << 16];
            int n;
            while ((n = in.read(buf)) > 0) bos.write(buf, 0, n);
            return bos.toByteArray();
        } catch (Throwable t) {
            return null;
        }
    }

    @Override
    public String modelName() {
        return "kokoro-82m-v1.0-q8f16";
    }

    @Override
    public String provider() {
        return provider;
    }

    /**
     * Load the model twice: the cold figure a listener pays on first use, and
     * the warm one that decides whether deck §5 item 6's "load at app start
     * and keep the session warm" mitigation is worth anything.
     */
    @Override
    public double[] load() {
        long t0 = System.nanoTime();
        session = makeSession();
        long t1 = System.nanoTime();
        OrtSession second = makeSession();
        long t2 = System.nanoTime();
        if (second != null) {
            try { second.close(); } catch (Throwable ignored) { }
        }
        return new double[]{ (t1 - t0) / 1e6, (t2 - t1) / 1e6 };
    }

    private OrtSession makeSession() {
        try {
            if (env == null) env = OrtEnvironment.getEnvironment();
            OrtSession.SessionOptions opts = new OrtSession.SessionOptions();
            opts.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT);
            provider = "cpu";
            return env.createSession(modelPath, opts);
        } catch (Throwable t) {
            Log.e(TAG, "could not open the Kokoro session", t);
            return null;
        }
    }

    /**
     * One line. Returns {@code [0, 0]} on any failure, which
     * {@code kokoro-probe.js} turns into an unmeasured RTF — and
     * {@code probeVerdict} FAILS an unmeasured RTF rather than passing it.
     */
    @Override
    public double[] synthesize(int[] ids, double speed) {
        if (session == null || ids == null || ids.length <= 2) return new double[]{0, 0};
        long t0 = System.nanoTime();
        int samples = run(ids, speed);
        long t1 = System.nanoTime();
        return new double[]{ (t1 - t0) / 1e6, samples / SAMPLE_RATE };
    }

    private int run(int[] ids, double speed) {
        Map<String, OnnxTensor> inputs = new HashMap<>();
        try {
            long[] tokens = new long[ids.length];
            for (int i = 0; i < ids.length; i++) tokens[i] = ids[i];
            inputs.put("input_ids", OnnxTensor.createTensor(
                    env, LongBuffer.wrap(tokens), new long[]{1, tokens.length}));

            /* THE STYLE ROW IS CHOSEN BY THE UNPADDED LENGTH. `ids` arrives
               with a pad at each end (tools/narration/kokoro-vocab.json
               documents the encoding), so the row is `length - 2`, clamped
               because a line longer than the matrix has no row of its own and
               the last row is the least wrong answer. K-04 chunks instead. */
            int row = Math.min(Math.max(ids.length - 2, 0), STYLE_ROWS - 1);
            float[] styleRow = new float[STYLE_DIM];
            System.arraycopy(style, row * STYLE_DIM, styleRow, 0, STYLE_DIM);
            inputs.put("style", OnnxTensor.createTensor(
                    env, FloatBuffer.wrap(styleRow), new long[]{1, STYLE_DIM}));

            inputs.put("speed", OnnxTensor.createTensor(
                    env, FloatBuffer.wrap(new float[]{(float) speed}), new long[]{1}));

            try (OrtSession.Result result = session.run(inputs)) {
                Object value = result.get(0).getValue();
                /* The samples are COUNTED AND DROPPED. K-01 measures speed,
                   memory and whether the passage survives a locked screen;
                   what it sounds like is K-03's audition, rendered on a
                   workstation from the same graph and the same weights.
                   Playing it here would mean an AudioTrack, which is how this
                   file would become the narration path by accident. */
                return countSamples(value);
            }
        } catch (Throwable t) {
            Log.e(TAG, "Kokoro inference failed", t);
            return 0;
        } finally {
            for (OnnxTensor t : inputs.values()) {
                try { t.close(); } catch (Throwable ignored) { }
            }
        }
    }

    /** The output is float[n] or float[1][n] depending on the export's leading
     *  dimension. Both are handled rather than one assumed: a re-export that
     *  added or dropped the batch axis would otherwise silently return 0
     *  samples, which reads as "infinitely fast" in an RTF. */
    private static int countSamples(Object value) {
        if (value instanceof float[]) return ((float[]) value).length;
        if (value instanceof float[][]) {
            float[][] rows = (float[][]) value;
            int n = 0;
            for (float[] r : rows) n += r.length;
            return n;
        }
        Log.e(TAG, "unexpected output type " + (value == null ? "null" : value.getClass()));
        return 0;
    }
}

File: mobile/plugins/foray-tts/ios/Sources/ForayTtsPlugin/KokoroOrtProbeEngine.swift
import Foundation
import OnnxRuntimeBindings
import os

/* K-01's runtime: ONNX Runtime, driven directly, three tensors in and one out.
 *
 * ── Why this file exists at all ───────────────────────────────────────────
 * K-01 landed the probe with its engine seam EMPTY and said so:
 * "`probeEngine` is nil on every build today. K-04 fills it." The founder then
 * ran the probe on build 2026091212 and got
 *
 *     voiceProbe kokoro-probe could not measure: passage-unphonemized
 *
 * — the first of four refusals, and behind it two more waiting: `model-absent`
 * (no weights in the bundle) and `engine-absent` (no runtime). Clearing only
 * the first would have moved the founder one refusal along and cost him a
 * second trip to a locked phone. All three are cleared together or the card
 * has not moved.
 *
 * ── Why ONNX Runtime and not sherpa-onnx / MLX ────────────────────────────
 * Deck §4, unchanged: sherpa-onnx's TTS API takes TEXT and phonemizes it
 * internally with espeak-ng, which is the GPL dependency this whole design
 * exists to keep off the phone; MLX Swift is iOS 18+ and this package's floor
 * is iOS 15. ORT is MIT, takes tensors, and has nothing to say about text.
 *
 * ── What this is NOT ──────────────────────────────────────────────────────
 * NOT a narration path. It has no `speak`, no audio session, no
 * `AVAudioEngine`, and nothing in `queue-manager.js` can reach it. It answers
 * `KokoroProbeEngine` — ids in, timings out — and K-04 is where a real engine
 * with chunking, a ring buffer and a `finished` event gets designed against
 * the numbers this produces. The samples are counted and thrown away: K-01
 * measures whether the phone CAN, not what it sounds like.
 *
 * ── The graph, documented ─────────────────────────────────────────────────
 * `onnx-community/Kokoro-82M-v1.0-ONNX`: `input_ids` (int64, [1, N] with a pad
 * at each end, N ≤ 512), `style` (float32, [1, 256]), `speed` (float32, [1]);
 * one float32 output at 24 kHz. The style row is chosen by the UNPADDED id
 * count — the voice file is 510 rows of 256 floats, one per possible length —
 * which is why `ids.count - 2` appears below and why dropping a single phoneme
 * would change the voice as well as the word.
 */

/// Whether this build can even try. `false` on any build that did not fetch
/// the weights; the plugin turns that into `model-absent` before it ever
/// constructs an engine, so a founder is told which artefact is missing.
enum KokoroModelFiles {
    /// The weights, looked for where the build step puts them.
    ///
    /// TWO PLACES, IN THIS ORDER, and the order is the finding: Capacitor's
    /// iOS template carries the web bundle as a FOLDER REFERENCE at
    /// `App.app/public`, so a file dropped into the generated
    /// `ios/App/App/public` after `cap sync` ships without any Xcode project
    /// surgery — that is what `tools/mobile/inject-models.mjs` does and it is
    /// the only injection point in the generated project that does not mean
    /// editing a `.pbxproj` from a Node script. The bundle root is tried
    /// FIRST anyway, so that if K-04 later adds a proper resource build phase
    /// this code needs no change to prefer it.
    static func modelURL() -> URL? {
        if let root = Bundle.main.url(forResource: ForayTtsPlugin.MODEL_RESOURCE,
                                      withExtension: ForayTtsPlugin.MODEL_EXTENSION) {
            return root
        }
        return Bundle.main.url(forResource: ForayTtsPlugin.MODEL_RESOURCE,
                               withExtension: ForayTtsPlugin.MODEL_EXTENSION,
                               subdirectory: ForayTtsPlugin.RESOURCE_SUBDIR)
    }

    /// The style matrix for the one voice the probe bundles.
    static func voiceURL() -> URL? {
        if let root = Bundle.main.url(forResource: ForayTtsPlugin.VOICE_RESOURCE, withExtension: "bin") {
            return root
        }
        return Bundle.main.url(forResource: ForayTtsPlugin.VOICE_RESOURCE,
                               withExtension: "bin",
                               subdirectory: ForayTtsPlugin.RESOURCE_SUBDIR)
    }
}

/// The probe's ONNX Runtime engine. One instance per probe run.
final class KokoroOrtProbeEngine: KokoroProbeEngine {
    /// Kokoro v1.0 emits 24 kHz. Hard-coded rather than read from the graph
    /// because the graph does not carry it — it is a property of the model
    /// card, and a wrong value here would scale every RTF by the ratio without
    /// changing anything a reader could see.
    static let SAMPLE_RATE: Double = 24_000

    /// The style matrix's shape, from the voice file's own length:
    /// 510 * 256 * 4 = 522,240 bytes, which is what
    /// `tools/mobile/fetch-models.mjs` pins and what its test asserts.
    static let STYLE_ROWS = 510
    static let STYLE_DIM = 256

    private static let log = Logger(subsystem: "ai.jwlabs.foura", category: "kokoro-probe")

    private let modelPath: String
    private let style: [Float]
    private var session: ORTSession?
    private var env: ORTEnv?
    private var loadColdMs: Double = 0
    private var loadWarmMs: Double = 0
    private var providerName = "cpu"

    let modelName = "kokoro-82m-v1.0-q8f16"
    var provider: String { providerName }

    /// `nil` when anything needed is absent or malformed. A FAILED
    /// CONSTRUCTION IS NOT A CRASH and not a zero: the plugin reports
    /// `engine-absent`, which is one of the four closed reason codes a founder
    /// reads off the screen.
    init?() {
        guard let model = KokoroModelFiles.modelURL(), let voice = KokoroModelFiles.voiceURL() else {
            return nil
        }
        guard let raw = try? Data(contentsOf: voice) else { return nil }
        let expected = Self.STYLE_ROWS * Self.STYLE_DIM * MemoryLayout<Float>.size
        guard raw.count == expected else {
            /* A voice file of the wrong length is a build that fetched
               something else. Refusing beats synthesizing with 256 floats read
               from the middle of an unrelated file, which would produce sound
               and therefore a number. */
            Self.log.error("voice file is \(raw.count) bytes, expected \(expected)")
            return nil
        }
        self.modelPath = model.path
        self.style = raw.withUnsafeBytes { buf in
            Array(buf.bindMemory(to: Float.self))
        }
    }

    /// Load the model twice. The first is the cold figure a listener pays on
    /// first use; the second is what warm re-entry costs, which is the number
    /// deck §5 item 6's "load at app start and keep the session warm"
    /// mitigation actually turns on.
    func load() -> (coldMs: Double, warmMs: Double) {
        loadColdMs = timed { self.session = self.makeSession() }
        loadWarmMs = timed { _ = self.makeSession() }
        return (loadColdMs, loadWarmMs)
    }

    /// ORT's environment, MADE ONCE AND KEPT. It used to be constructed inside
    /// `makeSession()`, which meant the warm load overwrote `self.env` and
    /// dropped the last strong reference the cold session's environment had —
    /// the session that is then kept and run on. ORT refcounts its global
    /// environment so this is survivable rather than fatal, but "survivable"
    /// is not a property to rely on underneath the one measurement this card
    /// exists to take, and one env for one engine is what the Android half
    /// already does (`OrtEnvironment.getEnvironment()` is a singleton there).
    private func environment() -> ORTEnv? {
        if let env { return env }
        env = try? ORTEnv(loggingLevel: ORTLoggingLevel.warning)
        return env
    }

    private func makeSession() -> ORTSession? {
        do {
            guard let env = environment() else { return nil }
            let options = try ORTSessionOptions()
            try options.setIntraOpNumThreads(0)   // 0 = ORT picks, per the C API
            try options.setGraphOptimizationLevel(ORTGraphOptimizationLevel.all)
            /* NO EXECUTION PROVIDER IS APPENDED, and `acceleratorWired` says so
               rather than leaving a reader to infer it from `provider: "cpu"`.
               ORT runs the CPU provider when nothing else is registered, so the
               `cpu` in #685's reading was NOT a CoreML attempt that fell back —
               CoreML was never wired at all. Wiring it means linking an ORT
               build that carries the CoreML EP and appending it here, which is
               a change to the one binary dependency `ios-shell` compiles and
               therefore its own card (K-08), not a line smuggled into a fix. */
            providerName = "cpu"
            return try ORTSession(env: env, modelPath: modelPath, sessionOptions: options)
        } catch {
            Self.log.error("could not open the Kokoro session: \(error.localizedDescription)")
            return nil
        }
    }

    /// No accelerator EP is registered on any build today. See `makeSession`.
    var acceleratorWired: Bool { false }

    /// One line. On failure the audio seconds are 0 AND the reason is named —
    /// `player/kokoro-probe.js`'s `SYNTH_REASONS`, which the plugin carries
    /// across as `detail`.
    ///
    /// THE REASON IS THE FIX FOR #685. This used to return a bare `(0, 0)` and
    /// claim, in this very comment, that `kokoro-probe.js` would read it as an
    /// unmeasured RTF. It did not: the plugin never put the rendered seconds on
    /// the wire, so the page divided a zero synthesis time by the passage's
    /// planning estimate and got `0.00` — a real number, better than every
    /// ceiling in the card. A failure has to arrive as a failure, by name.
    func synthesize(ids: [Int], speed: Double) -> (synthMs: Double, audioSec: Double, reason: String?) {
        guard let session else { return (0, 0, "session-absent") }
        guard ids.count > 2 else { return (0, 0, "zero-samples") }
        var samples = 0
        var failure: String? = nil
        let ms = timed {
            let out = self.run(session: session, ids: ids, speed: speed)
            samples = out.samples
            failure = out.reason
        }
        if samples <= 0 { return (ms, 0, failure ?? "zero-samples") }
        return (ms, Double(samples) / Self.SAMPLE_RATE, nil)
    }

    private func run(session: ORTSession, ids: [Int], speed: Double) -> (samples: Int, reason: String?) {
        do {
            /* int64, little-endian, exactly as the graph declares. `Int` is
               64-bit on every device this ships to, but the conversion is
               written out rather than assumed because a silent truncation here
               would be a different phoneme, not a crash. */
            var tokens = ids.map { Int64($0) }
            let idsData = NSMutableData(bytes: &tokens, length: tokens.count * MemoryLayout<Int64>.size)
            let idsValue = try ORTValue(tensorData: idsData,
                                        elementType: ORTTensorElementDataType.int64,
                                        shape: [1, NSNumber(value: tokens.count)])

            /* THE STYLE ROW IS CHOSEN BY THE UNPADDED LENGTH. `ids` arrives
               with a pad at each end (`tools/narration/kokoro-vocab.json`
               documents the encoding), so the row is `count - 2`, clamped
               because a line longer than the matrix has no row of its own and
               the last row is the least wrong answer. K-04 chunks instead. */
            let row = min(max(ids.count - 2, 0), Self.STYLE_ROWS - 1)
            var styleRow = Array(style[(row * Self.STYLE_DIM)..<((row + 1) * Self.STYLE_DIM)])
            let styleData = NSMutableData(bytes: &styleRow, length: styleRow.count * MemoryLayout<Float>.size)
            let styleValue = try ORTValue(tensorData: styleData,
                                          elementType: ORTTensorElementDataType.float,
                                          shape: [1, NSNumber(value: Self.STYLE_DIM)])

            var speedValue = Float(speed)
            let speedData = NSMutableData(bytes: &speedValue, length: MemoryLayout<Float>.size)
            let speedTensor = try ORTValue(tensorData: speedData,
                                           elementType: ORTTensorElementDataType.float,
                                           shape: [1])

            /* Output names come from the GRAPH, not from a constant. The
               onnx-community export calls it `waveform`; a re-export that
               renamed it would otherwise turn into "ORT returned nothing" with
               no diagnosis attached. */
            guard let outputName = (try session.outputNames()).first else { return (0, "no-output") }
            let outputs = try session.run(
                withInputs: ["input_ids": idsValue, "style": styleValue, "speed": speedTensor],
                outputNames: [outputName],
                runOptions: nil)
            /* `tensorDataWithError:` arrives in Swift as a throwing
               `tensorData()` — the ObjC-to-Swift error translation drops the
               `WithError:` suffix and the out-parameter. Spelled the ObjC way
               it does not compile. */
            guard let audio = outputs[outputName] else { return (0, "no-output") }
            let data = try audio.tensorData() as Data
            /* The samples are COUNTED AND DROPPED. K-01 measures speed, memory
               and whether the passage survives a locked screen; what it sounds
               like is K-03's audition, rendered on a workstation from the same
               graph and the same weights. Playing it here would mean an audio
               session, which would mean this file could become the narration
               path by accident. */
            let samples = data.count / MemoryLayout<Float>.size
            return (samples, samples > 0 ? nil : "zero-samples")
        } catch {
            /* THE LOG LINE IS NOT THE REPORT. This `os_log` is unreachable from
               the founder's phone; the returned code is what reaches the
               diagnostics record he pastes, and #685 is what it costs when only
               the unreachable half exists. */
            Self.log.error("Kokoro inference failed: \(error.localizedDescription)")
            return (0, "inference-threw")
        }
    }

    /// Wall-clock milliseconds around a block. `CACurrentMediaTime`-equivalent
    /// via a monotonic clock: `Date()` would be wrong across an NTP step, and a
    /// probe that runs for ten minutes on battery is exactly long enough for
    /// one.
    private func timed(_ body: () -> Void) -> Double {
        let start = DispatchTime.now().uptimeNanoseconds
        body()
        return Double(DispatchTime.now().uptimeNanoseconds - start) / 1_000_000
    }
}
